Add `wp core language update` command and Behat tests

wrnx opened the pull request:
Should fix #1387

danielbachhuber commented on php/WP_CLI/CommandWithTranslation.php:
```diff
@@ -125,6 +125,95 @@ public function install( $args, $assoc_args ) {
+
+		// Ignore updates for the default locale.
+		if ( 'en_US' == get_locale() ) {
+			\WP_CLI::line( "Translations updates are not needed for the default locale." );
```
Can we call this the `en_US` locale, just to be explicitly clear?

wrnx replied:
Yes. I think that "English (US)" would be even better.
```php
\WP_CLI::success( "Translations updates are not needed for the 'English (US)' locale." );
```

danielbachhuber commented on php/WP_CLI/CommandWithTranslation.php:
```diff
@@ -125,6 +125,95 @@ public function install( $args, $assoc_args ) {
+		wp_update_themes();  // Check for Theme updates.
+		wp_update_plugins(); // Check for Plugin updates.
+
+		$updates = wp_get_translation_updates(); // Retrieves a list of all translations updates available.
```
`wp_get_translation_updates()` doesn't perform the necessary updates?

wrnx replied:
It just gets a list of translation updates from the database.

The functions `wp_version_check`, `wp_update_themes` and `wp_update_plugins` checks for updates (which include translations updates) and store it in the database. After that, the function `wp_get_translation_updates` gets the info and return it as an array.

Translations updates are performed by the method `upgrade` (single update) or `bulk_upgrade` (all updates at once) of the WordPress class `Language_Pack_Upgrader`.

danielbachhuber replied:
Makes sense. Can you include some mention of "after the theme / plugin updates" have run?

danielbachhuber commented on php/WP_CLI/CommandWithTranslation.php:
```diff
@@ -125,6 +125,95 @@ public function install( $args, $assoc_args ) {
+		$updates = wp_get_translation_updates(); // Retrieves a list of all translations updates available.
+
+		if ( empty( $updates ) ) {
+			\WP_CLI::line( "No translations updates available." );
```
Let's follow the pattern of `wp core update` and use `WP_CLI::success( "Translations are up to date." );`

wrnx replied:
Done.

danielbachhuber commented on php/WP_CLI/CommandWithTranslation.php:
```diff
@@ -125,6 +125,95 @@ public function install( $args, $assoc_args ) {
+
+		// Formats the updates list.
+		foreach ( $updates as $update ) {
+			if ( 'plugin' == $update->type ) {
```
Is `type` always set?

wrnx replied:
Although not documented, it seems to be the default behavior of the WordPress API.

danielbachhuber commented on php/WP_CLI/CommandWithTranslation.php:
```diff
@@ -125,6 +125,95 @@ public function install( $args, $assoc_args ) {
+
+		if ( $num_to_update == $num_updated ) {
+			\WP_CLI::success( $line );
+		} else if ( $num_updated > 0 ) {
```
Should return a more descriptive error code if it was a partial failure? Related #1430

wrnx replied:
I didn't notice that the  ```WP_CLI::warning``` method doesn't return some error code.
I think that a simple behavior would be: ```WP_CLI::success``` exits with 0; ```WP_CLI::warning``` with 1; and ```WP_CLI::error``` with 2, since these methods may be used only for exit messages and maybe have a parameter to not do so.

I will put messages like "Translation updated successfully" or "Translation update failed" after each update to make it clear for the user what is going on.

danielbachhuber commented:
`--dry-run` is fine, but can we also have some indicator as to whether there's a new version in `wp core language list` (similar to `wp plugin list`)?

Also, I wasn't aware the plugin / theme translation API was complete. Is it? Can you link me to the Trac ticket?

wrnx commented:
Although not documented, it seems to be the default behavior of the WordPress.org API.
And the WordPress class `Language_Pack_Upgrader` relies on this to make the decision of which directory store the translation.

wrnx commented:
We can replace the 'updated' column, with the date, by an 'update' column, with the available/none message as the shown in the `wp plugin list` command.

About the translation API. I think it is since WordPress 4.0. I found these references:
[Internationalization goals for 4.0](https://make.wordpress.org/core/2014/05/21/internationalization-goals-for-4-0/)
[WordPress.org API - Translations](http://codex.wordpress.org/WordPress.org_API#Translations)
[Changeset 29630 – WordPress Trac](https://core.trac.wordpress.org/changeset/29630)
[Code Reference: translations_api](https://developer.wordpress.org/reference/functions/translations_api/)

wrnx commented:
The current messages, of the `wp core language update` command, are:
```
Updating 'English (UK)' translation for Akismet 3.1...
Downloading translation from https://downloads.wordpress.org/translation/plugin/akismet/3.1/en_GB.zip...
Translation updated successfully.
Success: Updated 1/1 translations.
```
If we use the ```bulk_upgrade``` method from the class ```Language_Pack_Upgrader``` (like the commands `wp plugin update --all` and `wp theme update --all` do), it will be like this:
```
Downloading translation from https://downloads.wordpress.org/translation/plugin/akismet/3.1/en_GB.zip...
Unpacking the update...
Installing the latest version...
Translation updated successfully.
Success: Updated 1/1 translations.
```
This will follow the pattern of the `update` commands, but we can't control the messages of the last one. So which one should we use?

Also, when we choose to update core/themes/plugins, the WordPress automatically tries to install translations updates, but it show the messages with HTML tags. I think it can't be changed. Unless we block it and call the `wp core language update` command after a core/themes/plugins update.

danielbachhuber commented on php/WP_CLI/CommandWithTranslation.php:
```diff
@@ -42,7 +42,7 @@
 	 * * native_name
 	 * * status
-	 * * updated
+	 * * update
```
I don't think we should drop the existing field. Can't they live side by side?

wrnx replied:
Yes, they can.

danielbachhuber commented on php/WP_CLI/CommandWithTranslation.php:
```diff
@@ -125,6 +140,112 @@ public function install( $args, $assoc_args ) {
+	 * : Preview which translations would be updated.
+	 *
+	 * [--force]
+	 * : Force the check for updates.
```
We should make `--force` the default behavior, and drop the argument. `wp (plugin|theme) list` both force refresh updates, which I think is an acceptable behavior.

wrnx replied:
Ok.

danielbachhuber commented on php/WP_CLI/CommandWithUpgrade.php:
```diff
@@ -10,6 +10,13 @@
 	protected $upgrade_transient;
 
+	function __construct() {
+		// Do not automatically check translations updates after updating plugins/themes.
```
Why?

wrnx replied:
The `Language_Pack_Upgrader::async_upgrade` method returns HTML tags to the standard output.
I corrected this by adding a call to the `wp core language update` command after the `remove_action` function:
```php
\WP_CLI::run_command( array( 'core', 'language', 'update' ), array( 'dry-run' => false ) );
```
You still need to run the command manually after the update, because after the plugin/theme update it will automatically check for translations updates to the older version of the plugin/theme instead of the new. Its the same behaviour as the default method, so its seems like a WordPress bug.

danielbachhuber replied:
Same as #1501? Can you comment on the core issue?

Seems pretty dirty to add a filter inside of `__construct()` — should this always run, or just on specific contexts?

danielbachhuber commented:
Can you make sure there's some test coverage for the change to `wp core language list` too? Looking pretty good, aside from some small nits.

danielbachhuber commented on php/WP_CLI/CommandWithTranslation.php:
```diff
@@ -125,6 +143,106 @@ public function install( $args, $assoc_args ) {
+
+		// Ignore updates for the default locale.
+		if ( 'en_US' == get_locale() ) {
+			\WP_CLI::success( "Translations updates are not needed for the 'English (US)' locale." );
```
I think this should be "Translation updates", note no "s" on "Translation"

wrnx replied:
I thought as the command deals with translations of core, plugins and themes, it would be "Translations" (plural).

danielbachhuber commented on php/WP_CLI/CommandWithTranslation.php:
```diff
@@ -125,6 +143,106 @@ public function install( $args, $assoc_args ) {
+		wp_clean_update_cache(); // Clear existing update caches.
+		wp_version_check();  // Check for Core updates.
+		wp_update_themes();  // Check for Theme updates.
+		wp_update_plugins(); // Check for Plugin updates.
```
Can you explicitly specify _translation_ updates?

wrnx replied:
I'm not sure about this, because these functions do not check only for translation updates.

danielbachhuber replied:
> I'm not sure about this, because these functions do not check only for translation updates.

The comment is for why the function calls are in this location though. And this location is specific to translations.

wrnx replied:
Alright then. Like this?
```php
wp_clean_update_cache(); // Clear existing update caches.
wp_version_check();      // Check for Core translation updates.
wp_update_themes();      // Check for Theme translation updates.
wp_update_plugins();     // Check for Plugin translation updates.
```

danielbachhuber replied:
:+1:

danielbachhuber commented:
Just some minor string updates, then this will be good to :ship:

danielbachhuber commented:
@wrnx Thanks for your work on this!

damirock commented:
Hi! Can't find language update in master version in my shell. Is it shipped?

danielbachhuber commented:
> Is it shipped?

Not yet, but shortly.
